cmd/nodeagent/nodeagentmanager.go

```diff
@@ -75,6 +82,76 @@ func (n *NodeAgentManager) startWatch() {
 	go informer.Run(n.watchStopper)
 }
 
+
+// runHeartbeatTimeCheckerLoop runs HeartbeatTimeChecker on an infinite loop
+func (n *NodeAgentManager) runHeartbeatTimeCheckerLoop() {
+	n.firstHeartbeatTimeout = ParseInt64FromEnv("FIRST_HEARTBEAT_TIMEOUT", 60000)
+	n.heartbeatTimeout = ParseInt64FromEnv("HEARTBEAT_TIMEOUT", 5000)
+	go func() {
+		for {
+			n.HeartbeatTimeChecker()
+			time.Sleep(time.Duration(n.heartbeatTimeout) * time.Millisecond)
+		}
+	}()
+}
+
+// HeartbeatTimeChecker checks that heartbeats are still being sent for each GameServerInfo
+// in the local gameServerMap, if not it will send a patch to mark those GameServers as unhealthy,
+// it follows these two rules:
+// 1. if the server hasn't sent its first heartbeat, it has FirstHeartbeatTimeout
+//    milliseconds since its creation before being marked as unhealthy
+// 2. if the server has sent its first heartbeat, it has HeartbeatTimeout milliseconds
+//    since its last heartbeat before being marked as unhealthy
+func (n *NodeAgentManager) HeartbeatTimeChecker() {
+	n.gameServerMap.Range(func(key interface{}, value interface{}) bool{
+		currentTime := n.nowFunc().UnixMilli()
+		gsd := value.(*GameServerInfo)
+		markUnhealthy := false
+		gsd.Mutex.RLock()
+		gameServerName := key.(string)
+		gameServerNamespace := gsd.GameServerNamespace
+		if gsd.LastHeartbeatTime == 0 && gsd.CreationTime != 0 &&
+		   (currentTime - gsd.CreationTime) > n.firstHeartbeatTimeout &&
+		   gsd.PreviousGameHealth != unhealthyStatus {
+			markUnhealthy = true
+		} else if gsd.LastHeartbeatTime != 0 &&
+				  (currentTime - gsd.LastHeartbeatTime) > n.heartbeatTimeout &&
+				  gsd.PreviousGameHealth != unhealthyStatus {
+			markUnhealthy = true
+		}
+		gsd.Mutex.RUnlock()
+		if markUnhealthy {
+			n.markGameServerUnhealthy(gameServerName, gameServerNamespace)
+		}
+		return true
+	})
+}
+
+// markGameServerUnhealthy sends a patch to mark the GameServer, described by its name
+// and namespace, as Unhealthy
+func (n *NodeAgentManager) markGameServerUnhealthy(gameServerName, gameServerNamespace string) {
+	logger := getLogger(gameServerName, gameServerNamespace)
+	logger.Infof("GameServer has not sent any heartbeats, marking Unhealthy")
```

Comment:
Can we have a different error message depending if the server became unhealthy for total lack of heartbeats or because of not receiving heartbeat for some time?